### DISCUSSION TITLE
How to perform inference with a model I trained with docTR?

### DISCUSSION THREAD

**Discussion by kforcodeai** [mindee/doctr#568] [Q&A] [answered]
I have trained a detection and recognition model on my dataset using your framework. 
But I am unable to use it using ocr_predictor.
How do I pass the model path to https://github.com/mindee/doctr/blob/b149266ea57fd59047193a01c328c2b8ecb9330a/doctr/models/zoo.py#L26
```Python
def ocr_predictor(
    det_arch: str = 'db_resnet50',
    reco_arch: str = 'crnn_vgg16_bn',
    pretrained: bool = False,
    **kwargs: Any
) -> OCRPredictor:
    """End-to-end OCR architecture using one model for localization, and another for text recognition.
    Example::
        >>> import numpy as np
        >>> from doctr.models import ocr_predictor
        >>> model = ocr_predictor('db_resnet50', 'crnn_vgg16_bn', pretrained=True)
        >>> input_page = (255 * np.random.rand(600, 800, 3)).astype(np.uint8)
        >>> out = model([input_page])
    Args:
        arch: name of the architecture to use ('db_sar_vgg', 'db_sar_resnet', 'db_crnn_vgg', 'db_crnn_resnet')
        pretrained: If True, returns a model pre-trained on our OCR dataset
    Returns:
        OCR predictor
    """

    return _predictor(det_arch, reco_arch, pretrained, **kwargs)
```
What kwargs should I use?
---

Also, even I put my models in cache_dir, the code goes on to download its own, also throws error for hashes unmatch
https://github.com/mindee/doctr/blob/b149266ea57fd59047193a01c328c2b8ecb9330a/doctr/models/data_utils.py

@fg-mindee

**Comment by fg-mindee** [comment, marked as answer]
Hello @k-for-code :wave: 

The factory function `ocr_predictor` is a bit more high-level than that, but you can easily achieve what you want :)
Here is a short example of to do this:

```python
import os

os.environ["USE_TORCH"] = "1"

import torch
from doctr.models.predictor import OCRPredictor
from doctr.models.detection.predictor import DetectionPredictor
from doctr.models.recognition.predictor import RecognitionPredictor
from doctr.models.preprocessor import PreProcessor
# from doctr.models.utils import load_pretrained_params

# Instantiate your model here
det_model = ...
reco_model = ...
# Load the checkpoints you produced
# load_pretrained_params(det_model, "<URL_TO_DET_CHECKPOINT>")
# load_pretrained_params(reco_model, "<URL_TO_RECO_CHECKPOINT>")
# If using PyTorch
# import torch
det_params = torch.load("path/to/your/local/det_checkpoint.pt", map_location="cpu")
reco_params = torch.load("path/to/your/local/reco_checkpoint.pt", map_location="cpu")
det_model.load_state_dict(det_params)
reco_model.load_state_dict(reco_params)

# Ask the preprocessor of each task to resize and normalize similarly to your training
# cf. https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L94 & https://github.com/mindee/doctr/blob/main/references/detection/train_pytorch.py#L109
det_predictor = DetectionPredictor(PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)
# cf. https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L97 & https://github.com/mindee/doctr/blob/main/references/recognition/train_pytorch.py#L111
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)

predictor = OCRPredictor(det_predictor, reco_predictor)
```

Let me know if you still have questions!

**Reply by kforcodeai** [reply]
@fg-mindee 
Please tell me am I doing it right
```Python
import os

os.environ["USE_TORCH"] = "1"

from doctr.models.predictor import OCRPredictor
from doctr.models.detection.predictor import DetectionPredictor
from doctr.models.preprocessor import PreProcessor
from doctr.models.utils import load_pretrained_params
from doctr.models import crnn_vgg16_bn, db_resnet50



det_model = crnn_vgg16_bn(pretrained=False)
rec_model = db_resnet50(pretrained=False)

load_pretrained_params(det_model, "/home/n13452/workSpace/docTR/my_model.pt")
```
load_pretrained_params calls 
https://github.com/mindee/doctr/blob/b27c3a664e99989e11d7bc527278abc33a539db0/doctr/models/data_utils.py#L46
```Python
def download_from_url(
    url: str,
    file_name: Optional[str] = None,
    hash_prefix: Optional[str] = None,
    cache_dir: Optional[str] = None,
    cache_subdir: Optional[str] = None,
) -> Path:
    """Download a file using its URL
    Example::
        >>> from doctr.models import download_from_url
        >>> download_from_url("https://yoursource.com/yourcheckpoint-yourhash.zip")
    Args:
        url: the URL of the file to download
        file_name: optional name of the file once downloaded
        hash_prefix: optional expected SHA256 hash of the file
        cache_dir: cache directory
        cache_subdir: subfolder to use in the cache
    Returns:
        the location of the downloaded file
    """

    if not isinstance(file_name, str):
        file_name = url.rpartition('/')[-1]

    if not isinstance(cache_dir, str):
        cache_dir = os.path.join(os.path.expanduser('~'), '.cache', 'doctr')

    # Check hash in file name
    if hash_prefix is None:
        r = HASH_REGEX.search(file_name)
        hash_prefix = r.group(1) if r else None

    folder_path = Path(cache_dir) if cache_subdir is None else Path(cache_dir, cache_subdir)
    file_path = folder_path.joinpath(file_name)
    # Check file existence
    if file_path.is_file() and (hash_prefix is None or _check_integrity(file_path, hash_prefix)):
        logging.info(f"Using downloaded & verified file: {file_path}")
        return file_path

    # Create folder hierarchy
    folder_path.mkdir(parents=True, exist_ok=True)
    # Download the file
    try:
        print(f"Downloading {url} to {file_path}")
        _urlretrieve(url, file_path)
    except (urllib.error.URLError, IOError) as e:  # type: ignore[attr-defined]
        if url[:5] == 'https':
            url = url.replace('https:', 'http:')
            print('Failed download. Trying https -> http instead.'
                  f" Downloading {url} to {file_path}")
            _urlretrieve(url, file_path)
        else:
            raise e

    # Remove corrupted files
    if isinstance(hash_prefix, str) and not _check_integrity(file_path, hash_prefix):
        # Remove file
        os.remove(file_path)
        raise ValueError(f"corrupted download, the hash of {url} does not match its expected value")

    return file_path
```
Now download_from_url function calls  _urlretrieve() , 
```
def _urlretrieve(url: str, filename: Union[Path, str], chunk_size: int = 1024) -> None:
    with open(filename, "wb") as fh:
        with urllib.request.urlopen(urllib.request.Request(url, headers={"User-Agent": USER_AGENT})) as response:
            with tqdm(total=response.length) as pbar:
                for chunk in iter(lambda: response.read(chunk_size), ""):
                    if not chunk:
                        break
                    pbar.update(chunk_size)
                    fh.write(chunk)
```
which call urllib.request.urlopen, which in turn  returns `unknown url type` ValueError
```
Exception has occurred: ValueError
unknown url type: '/home/n13452/workSpace/docTR/my_model.pt'
  File "/home/n13452/workSpace/docTR/doctr-main/doctr/models/data_utils.py", line 30, in _urlretrieve
    with urllib.request.urlopen(urllib.request.Request(url, headers={"User-Agent": USER_AGENT})) as response:
  File "/home/n13452/workSpace/docTR/doctr-main/doctr/models/data_utils.py", line 93, in download_from_url
    _urlretrieve(url, file_path)
  File "/home/n13452/workSpace/docTR/doctr-main/doctr/models/utils/pytorch.py", line 40, in load_pretrained_params
    archive_path = download_from_url(url, hash_prefix=hash_prefix, cache_subdir='models', **kwargs)
  File "/home/n13452/workSpace/docTR/doctr-main/script.py", line 16, in <module>
    load_pretrained_params(det_model, "/home/n13452/workSpace/docTR/my_model.pt")
```
I feel its a bug, please let me know If I am making any mistakes

**Reply by fg-mindee** [reply]
My bad, I just edited my previous answer!

`load_pretrained_params` takes a URL, if you have the checkpoint locally, you can use the `load_state_dict` method of the model :)
Let me know if you're still having issues

**Reply by kforcodeai** [reply]
@fg-mindee I feel I am lacking an understanding of the source code. 
Can you please please help with me some resources to learn, 
How to code at the level you do ?  
I am an intermediate-level developer, but your code takes me time to understand. How do I upgrade myself? 
Any suggestion, roadmap, resources will be of great help. I wanna know the secret

**Reply by fg-mindee** [reply]
Well first, did it answer your question? Did you manage to get it to work? :sweat_smile: 

For the rest, contributing to open source projects has usually been a good way to improve actually because you get feedback from peers!

**Reply by kforcodeai** [reply]
If i instantiate the model using the below code 
```Python
from doctr.models import sar_resnet31, db_resnet50
rec_model = sar_resnet31(pretrained=False)
```
it fails
![doctr](https://user-images.githubusercontent.com/59497032/140459220-ca0bcc0f-9ece-46b2-a77b-46d0afcf9f23.png)

**Reply by fg-mindee** [reply]
Mmmh that's strange. My best guess is that you're running the last release and not the dev build

Could you run this snippet and report back the result to confirm this please?
```shell
wget https://raw.githubusercontent.com/mindee/doctr/main/scripts/collect_env.py
python collect_env.py
```

**Reply by kforcodeai** [reply]
DocTR version: 0.4.1a0
TensorFlow version: 2.6.0
PyTorch version: 1.9.0+cu102 (torchvision 0.10.0+cu102)
OpenCV version: 4.4.0
OS: Ubuntu 20.04.3 LTS
Python version: 3.8.10
Is CUDA available (TensorFlow): No
Is CUDA available (PyTorch): No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

**Reply by fg-mindee** [reply]
Oh I see, your environment should be working perfectly. But it's because you didn't switch to PyTorch mode correctly :+1: 

replace 
```
%env "USE_TORCH" "1"
```
by this
```
os.environ["USE_TORCH"] = "1"
```
and that should do the trick ;)

**Reply by fg-mindee** [reply]
Did it solve your issue @k-for-code? :)

**Reply by kforcodeai** [reply]
# No
and please check you commented out 
```
os.environ["USE_TORCH"] = "1"
```
and added 
```
%env "USE_TORCH" "1"
```
in last reply code

The issue remains the same
![bug6](https://user-images.githubusercontent.com/59497032/140874668-723312ed-78a7-40ed-b2f7-39dc6e354678.png)

**Reply by fg-mindee** [reply]
This is really strange: I'm unable to reproduce this behaviour on my end. So it's quite hard to understand where that would come from.  I'm almost certain it's because your instantiated `sar_resnet31` is the TF version.
But the `os.environ["USE_TORCH"] = "1"` should have done the trick.

If you are running this in a notebook or something similar, make sure to execute
```python
import os
os.environ["USE_TORCH"] = "1"
```

before any import of doctr. To confirm that this has worked, at the end of your script, could you add the following line and report back the console output please?

```python
from doctr.file_utils import is_tf_available, is_torch_available
print(f"TF available: {is_tf_available()}, PyTorch available: {is_torch_available()}")
```
_you should be getting (False, True) for the snippet to work. And I suspect that you're having (True, True) or (True, False)_

**Reply by kforcodeai** [reply]
@fg-mindee, yes  I am using a Jupyter Notebook. 

![bug6](https://user-images.githubusercontent.com/59497032/141240413-d18d0bc7-a8e8-4bb0-ba31-4918e2ef9447.png)

**Reply by fg-mindee** [reply]
I'm really unable to reproduce this behaviour. Could you try to run your code with a python script or directly in command, rather than on Jupyter please?

**Reply by kforcodeai** [reply]
@fg-mindee  I was able to resolve if we move the import torch statement before importing sar_resent it works

**Reply by fg-mindee** [reply]
That's quite strange :sweat_smile: 

But glad you managed to solve this!

**Comment by adesgautam** [comment]
Hi, how to perform inference on trained recognition models using tensorflow ?

**Reply by fg-mindee** [reply]
Almost the same but only doing text recognition:
```python
import os

os.environ["USE_TF"] = "1"

from doctr.models.recognition.predictor import RecognitionPredictor
from doctr.models.preprocessor import PreProcessor

# Instantiate your model here
reco_model = ...
# Load the checkpoints you produced
reco_model.load_weights("path/to/your/local/reco_checkpoint")

# Ask the preprocessor of each task to resize and normalize similarly to your training
# cf. https://github.com/mindee/doctr/blob/main/references/recognition/train_tensorflow.py
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)

# Do inference
input_tensor = ...
out = reco_predictor(input_tensor)
```

Let me know if you encounter some problems!

**Reply by adesgautam** [reply]
Thanks you so much!

Used the following to load:

```
from doctr.models.recognition.predictor import RecognitionPredictor
from doctr.models.preprocessor import PreProcessor

model = crnn_vgg16_bn(pretrained=False, pretrained_backbone=False)
model.load_weights("doctr/crnn_vgg16_bn_20211214-090713/weights")
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), 
                                                   preserve_aspect_ratio=True, 
                                                   batch_size=1, 
                                                   mean=(0.694, 0.695, 0.693), 
                                                   std=(0.299, 0.296, 0.301)), 
                                      model)
```

**Comment by dhea1323** [comment]
Hi @fg-mindee ,
I want to make predictions separately between detection and recognition stage without ```predictor = OCRPredictor(det_predictor, reco_predictor)```, here's the example 

```python
doc = DocumentFile.from_pdf("path/to/pdf_file.pdf")
# Detection model
det_model = db_resnet50(pretrained=False)
det_param = torch.load("./path/to/load_model.pt", map_location="cpu")
det_model.load_state_dict(det_param)
det_predictor = DetectionPredictor(PreProcessor((1024, 1024), batch_size=1, mean=(0.798, 0.785, 0.772), std=(0.264, 0.2749, 0.287)), det_model)
detection = det_predictor(doc)

#Recognition model
reco_model = crnn_vgg16_bn(pretrained=False)
reco_param = torch.load("./path/to/load_model.pt", map_location="cpu")
reco_model.load_state_dict(reco_param)
reco_predictor = RecognitionPredictor(PreProcessor((32, 128), preserve_aspect_ratio=True, batch_size=32, mean=(0.694, 0.695, 0.693), std=(0.299, 0.296, 0.301)), reco_model)        
recognition = reco_predictor(detection)
```
and of course I got an error :sweat_smile:  because I don't know what supposed to do after getting the output from detection stage ```detection = det_predictor(doc)``` then proceed to the recognition stage ```recognition = reco_predictor(detection)```.

Here's the error : ```ValueError: incorrect input shape: all crops are expected to be multi-channel 2D images.```
I expect the final result to be a string with ```.render()```

Thank You.

**Reply by frgfm** [reply]
Oops, for some reasons I didn't see your message @dhea1323!

Did you manage to solve your problem?

**Comment by shahdghorsi** [comment]
Hello @fg-mindee ,
I used one of the pretrained models, I am trying to save the output instead of showing it, but I could not really achieve that.  
result = model(doc)
result.show(doc) # what to use instead of show() here to save the output.

Thanks,

**Reply by felixdittrich92** [reply]
Hi @adesgautam :wave:,

you have some options:

.render()  -> get the predicted text as String
.export() -> get results as json format dict
.export_as_xml() -> get the xml as string and the whole ElementTree object
.synthesize() -> get a list of each page (as blank image where the results are printed)   # i think you are searching for this :)

**Reply by shahdghorsi** [reply]
Thank you so much @felixdittrich92 that saved a picture.

**Reply by shahdghorsi** [reply]
I was wondering if it is possible to have the document as it is instead of an image, I do want to get it with the bounding boxes.

**Reply by felixdittrich92** [reply]
we do not provide a 'ready to use' functionality for this but you can solve it on your own :)

Ref: https://github.com/mindee/doctr/discussions/570#discussioncomment-1599113

you could for example break it down into a for loop and draw with opencv all boxes per page and save the image
